atd-toolbox/s3_restore_cr3_pdfs/s3_restore_cr3_pdfs.py

```diff
@@ -0,0 +1,268 @@
+            required=True,
+            metavar = 'crashes.json')
+
+    argparse.add_argument("--i-understand",
+            help = 'Do not ask the user to acknoledge that this program changes the state of S3 objects and the database.',
+            action = 'store_true')
+
+    args = argparse.parse_args()
+except:
+    # a stderr log is not needed, argparse croaks verbosly
+    sys.exit(1)
+
+
+# This program will change the state of S3 objects.  Make sure the user is OK with what is about to happen.
+try:
+    if not args.i_understand:
+        print('')
+        print("Warning: This program changes S3 Objects and the Vision Zero database.")
+        print('')
+        print("This program will restore previous PDF CR3 versions for crashes specified in the JSON object you provide.")
+        print("If there is a 'application/pdf' stored as a previous version of a specified crash's CR3, this program will restore that file to the current version.")
+        print("This program will also update the databases cr3_file_metadata field for crash records based on the S3 file restored.")
+        print("Please type 'I understand' to continue.")
+        print('')
+        ack = input()
+        assert(re.match("^i understand$", ack, re.I))
+except:
+    eprint("User acknoledgement failed.")
+    sys.exit(1)
+
+
+# verify that environment AWS variables were available and have populated values to be used to auth to AWS
+try:
+    assert(ACCESS_KEY is not None and SECRET_KEY is not None)
+except:
+    eprint("Please set environment variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY")
+    sys.exit(1)
+
+
+# verify that environment Hasura variables were available and have populated values
+try:
+    assert(HASURA_ADMIN_KEY is not None and HASURA_ENDPOINT is not None)
+except:
+    eprint("Please set environment variables HASURA_ADMIN_KEY and HASURA_ENDPOINT")
+    sys.exit(1)
+
+
+# sanity check the provided hasura endpoint and the assertion from the user about staging/production
+if (args.production):
+    try:
+        assert(not re.search("staging", HASURA_ENDPOINT, re.I))
+    except:
+        eprint("Production flag used but staging appears in the Hasura endpoint URL")
+        sys.exit(1)
+else:
+    try:
+        assert(re.search("staging", HASURA_ENDPOINT, re.I))
+    except:
+        eprint("Production flag is not used and staging doesn't appear in the Hasura endpoint URL")
+        sys.exit(1)
+
+
+# connect to AWS/S3 and validate connection
+try:
+    aws_config = Config(
+            region_name = 'us-east-1',
+            )
+
+    s3 = boto3.client(
+            's3', 
+            aws_access_key_id = ACCESS_KEY, 
+            aws_secret_access_key = SECRET_KEY, 
+            config = aws_config
+            )
+    prefix = ('production' if args.production else 'staging') + '/cris-cr3-files/'
+    s3.list_objects(Bucket = bucket, Prefix = prefix)
+except:
```

Comment:
@frankhereford the tract/catch pattern is good. for debugging, it's probably worthwhile to print the actual exception message before you exit. You can extend your handler like:

```python
try:
    # do stuff
except Exception as e:
    eprint("Unable to complete call to S3; check AWS credentials")
    # print string representation of Exception msg
    eprint(str(e))
    sys.exit(1)
```

Reply:
Got it - thanks! I'll do that for the places where that pattern is used.